File: clients/java/src/main/java/com/priostack/acn/FetchResult.java
package com.priostack.acn;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.ArrayList;
import java.util.List;

/** Result of {@link ACNClient#fetch} — the content reader. */
public final class FetchResult {

    private final String space;
    private final List<JsonNode> objects;
    private final int matched;
    private final int total;
    private final int returnedTokens;
    private final JsonNode raw;

    public FetchResult(
            String space,
            List<JsonNode> objects,
            int matched,
            int total,
            int returnedTokens,
            JsonNode raw) {
        this.space = space;
        this.objects = objects;
        this.matched = matched;
        this.total = total;
        this.returnedTokens = returnedTokens;
        this.raw = raw;
    }

    public String space() {
        return space;
    }

    /** The returned objects, each {@code {objectRef, kind, content, tokens}}. */
    public List<JsonNode> objects() {
        return objects;
    }

    public int matched() {
        return matched;
    }

    public int total() {
        return total;
    }

    public int returnedTokens() {
        return returnedTokens;
    }

    public JsonNode raw() {
        return raw;
    }

    /** Just the {@code content} strings of the returned objects. */
    public List<String> contents() {
        List<String> out = new ArrayList<>(objects.size());
        for (JsonNode o : objects) {
            out.add(o.path("content").asText(""));
        }
        return out;
    }

    @Override
    public String toString() {
        return "FetchResult{space="
                + space
                + ", matched="
                + matched
                + ", total="
                + total
                + ", returnedTokens="
                + returnedTokens
                + '}';
    }
}

File: clients/php/src/Result/FetchResult.php
<?php

declare(strict_types=1);

namespace Priostack\Acn\Result;

/**
 * Result of {@see \Priostack\Acn\Client::fetch()} (the content reader).
 *
 * Each entry in {@see $objects} is an associative array shaped like
 * {@code {objectRef, kind, content, tokens}}.
 */
final class FetchResult
{
    /**
     * @param list<array<string, mixed>> $objects
     * @param array<string, mixed>       $raw
     */
    public function __construct(
        public readonly string $space,
        public readonly array $objects,
        public readonly int $matched,
        public readonly int $total,
        public readonly int $returnedTokens,
        public readonly array $raw = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromData(array $data, string $fallbackSpace = ''): self
    {
        return new self(
            space: (string) ($data['space'] ?? $fallbackSpace),
            objects: array_values((array) ($data['objects'] ?? [])),
            matched: (int) ($data['matched'] ?? 0),
            total: (int) ($data['total'] ?? 0),
            returnedTokens: (int) ($data['returnedTokens'] ?? 0),
            raw: $data,
        );
    }

    /**
     * Just the {@code content} strings of the returned objects.
     *
     * @return list<string>
     */
    public function contents(): array
    {
        return array_map(
            static fn (array $o): string => (string) ($o['content'] ?? ''),
            $this->objects,
        );
    }
}
